src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ActiveDirectoryAuthenticationProvider.cs

```diff
@@ -320,5 +297,168 @@ private class CustomWebUi : ICustomWebUi
+                .WithAuthority(publicClientAppKey._authority)
+                .WithClientName(Common.DbConnectionStringDefaults.ApplicationName)
+                .WithClientVersion(Common.ADP.GetAssemblyVersion().ToString())
+                .WithRedirectUri(publicClientAppKey._redirectUri)
+                .WithParentActivityOrWindow(_parentActivityOrWindowFunc)
+                .Build();
+            }
+#endif
+#if NETFRAMEWORK
+            if (_iWin32WindowFunc != null)
+            {
+                publicClientApplication = PublicClientApplicationBuilder.Create(publicClientAppKey._applicationClientId)
+                .WithAuthority(publicClientAppKey._authority)
+                .WithClientName(Common.DbConnectionStringDefaults.ApplicationName)
+                .WithClientVersion(Common.ADP.GetAssemblyVersion().ToString())
+                .WithRedirectUri(publicClientAppKey._redirectUri)
+                .WithParentActivityOrWindow(_iWin32WindowFunc)
+                .Build();
+            }
+#endif
+#if !NETCOREAPP
+            else
+#endif
+            {
+                publicClientApplication = PublicClientApplicationBuilder.Create(publicClientAppKey._applicationClientId)
+                .WithAuthority(publicClientAppKey._authority)
+                .WithClientName(Common.DbConnectionStringDefaults.ApplicationName)
+                .WithClientVersion(Common.ADP.GetAssemblyVersion().ToString())
+                .WithRedirectUri(publicClientAppKey._redirectUri)
+                .Build();
+            }
+
+            return publicClientApplication;
+        }
+
+        internal class PublicClientAppKey
+        {
+            public readonly string _authority;
+            public readonly string _redirectUri;
+            public readonly string _applicationClientId;
+#if NETFRAMEWORK
+            public readonly Func<System.Windows.Forms.IWin32Window> _iWin32WindowFunc;
+#endif
+#if NETSTANDARD
+            public readonly Func<object> _parentActivityOrWindowFunc;
+#endif
+            private int _hashValue;
+
+            public PublicClientAppKey(string authority, string redirectUri, string applicationClientId
+#if NETFRAMEWORK
+            , Func<System.Windows.Forms.IWin32Window> iWin32WindowFunc
+#endif
+#if NETSTANDARD
+            , Func<object> parentActivityOrWindowFunc
+#endif
+                )
+            {
+                _authority = authority;
+                _redirectUri = redirectUri;
+                _applicationClientId = applicationClientId;
+#if NETFRAMEWORK
+                _iWin32WindowFunc = iWin32WindowFunc;
+#endif
+#if NETSTANDARD
+                _parentActivityOrWindowFunc = parentActivityOrWindowFunc;
+#endif
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (obj == null)
+                {
+                    return false;
+                }
+
+                PublicClientAppKey pcaKey = obj as PublicClientAppKey;
```

Comment:
If this conversion fails, a null ref exception will be thrown below. This check should be sufficient.

```
if (!(obj is PublicClientAppKey pcaKey))
{
    return false;
}
```